packages/solid/src/components/shared.tsx

```diff
@@ -0,0 +1,77 @@
+import type { ExpandedRouteConfig } from "@uploadthing/shared";
+
+export const generatePermittedFileTypes = (config?: ExpandedRouteConfig) => {
+  const fileTypes = config ? Object.keys(config) : [];
+
+  const maxFileCount = config
+    ? Object.values(config).map((v) => v.maxFileCount)
+    : [];
+
+  return { fileTypes, multiple: maxFileCount.some((v) => v && v > 1) };
+};
+
+export const capitalizeStart = (str: string) => {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+};
+
+export const INTERNAL_doFormatting = (config?: ExpandedRouteConfig): string => {
+  if (!config) return "";
+
+  const allowedTypes = Object.keys(config) as (keyof ExpandedRouteConfig)[];
+
+  const formattedTypes = allowedTypes.map((f) => (f === "blob" ? "file" : f));
+
+  // Format multi-type uploader label as "Supports videos, images and files";
+  if (formattedTypes.length > 1) {
+    const lastType = formattedTypes.pop();
+    return `${formattedTypes.join("s, ")} and ${lastType}s`;
+  }
+
+  // Single type uploader label
+  const key = allowedTypes[0];
+  const formattedKey = formattedTypes[0];
+
+  const { maxFileSize, maxFileCount } = config[key]!;
+
+  if (maxFileCount && maxFileCount > 1) {
+    return `${formattedKey}s up to ${maxFileSize}, max ${maxFileCount}`;
+  } else {
+    return `${formattedKey} (${maxFileSize})`;
+  }
+};
+
+export const allowedContentTextLabelGenerator = (
+  config?: ExpandedRouteConfig,
+): string => {
+  return capitalizeStart(INTERNAL_doFormatting(config));
+};
+
+export const progressWidths: Record<number, string> = {
+  0: "after:w-0",
+  10: "after:w-[10%]",
+  20: "after:w-[20%]",
+  30: "after:w-[30%]",
+  40: "after:w-[40%]",
+  50: "after:w-[50%]",
+  60: "after:w-[60%]",
+  70: "after:w-[70%]",
+  80: "after:w-[80%]",
+  90: "after:w-[90%]",
+  100: "after:w-[100%]",
+};
```

Comment:
all of this looks like a 1:1 copy of the react package? can we put it in `uploadthing/client` maybe and export with an `@internal` jsdoc note?

Reply:
Doesn't it make sense to move it to `@uploadthing/shared` then? I mean, to me it seems this package is exactly for shared stuff

Reply:
Yea but there's declaration support issues with the shared package (see https://github.com/pingdotgg/uploadthing/pull/272). But might be better to put it in shared first and then we'll move it later if we decide we have to

Reply:
Hm, if there are issues with the shared package, then I would prefer going with more robust option. In addition, I can see that we already have several utility functions inside of `uploadthing/client`. Probably, I should put style-related stuff in separate file and just re-export it from `client.ts` to not overload single file. What do you think?
<img width="749" alt="image" src="https://github.com/pingdotgg/uploadthing/assets/55754540/aaeeee9c-9e3b-402b-8520-704b6fb5ac86">

Reply:
>  I should put style-related stuff in separate file and just re-export it from client.ts to not overload single file. What do you think?

yea that sounds reasonable